docs/proposals/pod-security-context.md

```diff
@@ -0,0 +1,233 @@
+container, it requires more complex validations and allows for transcription errors between
+containers.
+
+### Use Case: Override pod security context for container
+
+Some use cases require the containers in a pod to run with different security settings.  As an
+example, a user may want to have a pod with two containers, one of which runs as root with the
+privileged setting, and one that runs as a non-root UID.  To support use cases like this, it should
+be possible to override appropriate (ie, not intrinsically pod-level) security settings for
+individual containers.
+
+## Analysis
+
+### SELinux context: pod- or container- level?
+
+Currently, SELinux context is specifiable only at the container level.  This is an inconvenient
+factoring for sharing volumes and other SELinux-secured resources between containers because there
+is no way in SELinux to share resources between processes with different MCS labels except to
+remove MCS labels from the shared resource.  This is a big security risk: _any container_ in the
+system can work with a resource which has the same SELinux context as it and no MCS labels.  Since
+we are also not interested in isolating containers in a pod from one another, the SELinux context
+should be shared by all containers in a pod to facilitate isolation from the containers in other
+pods and sharing resources amongst all the containers of a pod.
+
+### Pod-level supplemental groups
+
+The [generalized non-root and SELinux support for volumes](https://github.com/kubernetes/kubernetes/pull/12944)
+pull request describes in detail the cases for sharing volumes using supplemental groups.  The
+easiest way to share resources among containers running as different UID/primary GID combinations is
+to make them owned by a supplemental group that all containers run as.  Therefore, there should be a
+pod-level supplemental group field that cannot be overridden in containers.
+
+## Proposed Design
+
+The `SecurityContext` type should be replaced by two new, similar types: `PodSecurityContext` and
+`ContainerSecurityContext`.
+
+### PodSecurityContext
+
+The `PodSecurityContext` type expresses pod-level security settings and the base container-level
+settings.  
+
+```go
+type PodSpec struct {
+    // retained for compatibility with existing API
+    HostNetwork bool
+
+    // other fields omitted
+    SecurityContext *PodSecurityContext
+}
+
+type PodSecurityContext struct {
+    // Uses the host's network namespace. If this option is set, the ports that will be
+    // used must be specified.
+    // Optional: Default to false.
+    HostNetwork bool
+
+    // Uses the host's IPC namespace proposed in (https://github.com/kubernetes/kubernetes/pull/12470)
+    HostIPC bool
+
+    // The supplemental group ID that will own volumes in this pod and that all containers will run
+    // under as a supplemental group
+    SupplementalGroupID *int64
```

Comment:
to be clear, the model I assumed for volumes was that kubelet would just assign arbitrary GIDs to volumes in a pod and add that GID to all containers in a pod.  This is not something users generally need control over for things like emptyDir. For volumes that have data in them, its a bit of a different story